File: java/sliding_window/longest_repeating_character_replacement.java
class Solution 
{
    public int characterReplacement(String s, int k) 
    {
        // HashMap<Character, Integer> hash = new HashMap<>();
        int maxF    = 1;
        int res     = 0;
        int l       = 0;
        int r       = 0;
        int[] hash = new int[26];

        while ( r < s.length() )
        {
            int index = (int)s.charAt(r)-(int)'A';
            hash[index] += 1;
            maxF = Math.max(maxF, hash[index]);

            if ( r - l + 1 - maxF > k )
            {
                hash[(int)s.charAt(l)-(int)'A'] -= 1;
                l++;
            }
            res = Math.max(res, r - l + 1);
            r++;
        }
        // System.out.println("l "+l+" r "+r);

        return res;
    }
}

File: python/sliding_window/longest_repeating_character_replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        if len(s) == 0: return 0 
        l = 0
        freq = {}
        max_freq = 1
        res = 1

        for r in range(len(s)):
            if s[r] in freq:
                freq[s[r]] += 1
                max_freq = max(max_freq, freq[s[r]])
            else:
                freq[s[r]] = 1
            
            cur_len = r - l + 1
            while cur_len - max_freq > k:
                freq[s[l]] -= 1
                l += 1
                cur_len = r - l + 1

            res = max(cur_len, res)
        return res